crates/wiggle/src/lib.rs

```diff
@@ -545,33 +558,35 @@ impl<'a, T> GuestPtr<'a, [T]> {
             T::validate(unsafe { ptr.add(offs as usize) })?;
         }
 
-        // SAFETY: iff there are no overlapping mut borrows it is valid to construct a &[T]
+        // SAFETY: iff there are no overlapping mut borrows it is valid to
+        // construct a &[T].
         let ptr = unsafe { slice::from_raw_parts(ptr, self.pointer.1 as usize) };
-
-        Ok(GuestSlice {
+        Ok(Some(GuestSlice {
             ptr,
             mem: self.mem,
             borrow,
-        })
+        }))
     }
 
-    /// Attempts to create a [`GuestSliceMut<'_, T>`] from this pointer, performing
-    /// bounds checks and type validation. The `GuestSliceMut` is a smart pointer
-    /// that can be used as a `&[T]` or a `&mut [T]` via the `Deref` and `DerefMut`
-    /// traits. The region of memory backing the slice will be marked as borrowed
-    /// by the [`GuestMemory`] until the `GuestSlice` is dropped.
+    /// Attempts to create a [`GuestSliceMut<'_, T>`] from this pointer,
+    /// performing bounds checks and type validation. The `GuestSliceMut` is a
+    /// smart pointer that can be used as a `&[T]` or a `&mut [T]` via the
+    /// `Deref` and `DerefMut` traits. The region of memory backing the slice
+    /// will be marked as borrowed by the [`GuestMemory`] until the `GuestSlice`
+    /// is dropped.
     ///
-    /// This function will return a `GuestSliceMut` into host memory if all checks
-    /// succeed (valid utf-8, valid pointers, memory is not borrowed, etc). If
-    /// any checks fail then `GuestError` will be returned.
-    pub fn as_slice_mut(&self) -> Result<GuestSliceMut<'a, T>, GuestError>
+    /// This function will return a `GuestSliceMut` into host memory if all
+    /// checks succeed (valid utf-8, valid pointers, memory is not borrowed,
+    /// etc). If any checks fail then `GuestError` will be returned.
+    pub fn as_slice_mut(&self) -> Result<Option<GuestSliceMut<'a, T>>, GuestError>
     where
         T: GuestTypeTransparent<'a>,
     {
-        let len = match self.pointer.1.checked_mul(T::guest_size()) {
-            Some(l) => l,
-            None => return Err(GuestError::PtrOverflow),
-        };
+        if self.mem.is_shared_memory() {
+            return Ok(None);
+        }
```

Comment:
Similar requests as above for doc updates and comments on the test for shared memory here